torchtitan/grpo/grpo_step.py

```diff
@@ -175,6 +175,91 @@ def compose_grpo_loss(
 
     return loss
 
+def compute_token_entropy(pred, mask):
+    """
+    Compute entropy at each token position for identifying forking tokens.
+    
+    Args:
+        pred: Model predictions [batch_size, seq_len, vocab_size]
+        mask: Attention mask [batch_size, seq_len]
+    
+    Returns:
+        entropy: Entropy at each position [batch_size, seq_len]
+    """
+    log_probs = torch.nn.functional.log_softmax(pred, dim=-1)
+    probs = torch.exp(log_probs)
+    entropy = -(probs * log_probs).sum(dim=-1)
+    return entropy * mask
+
+
+def apply_simko_adjustment(
+    ratio, pred, labels, reward, mask,
+    alpha=0.01, K=3, lambda_top1=1.1, tau_percentile=80
+):
+    """
+    Apply SimKO adjustments to policy ratio.
+    
+    Args:
+        ratio: Original policy ratio [batch_size, seq_len]
+        pred: Model predictions [batch_size, seq_len, vocab_size]
+        labels: Target labels [batch_size, seq_len]
+        reward: Advantages [batch_size, seq_len]
+        mask: Attention mask [batch_size, seq_len]
+        alpha: Label smoothing strength
+        K: Number of top candidates
+        lambda_top1: Penalty multiplier for top-1 negatives
+        tau_percentile: Entropy threshold percentile
+    
+    Returns:
+        adjusted_ratio: Ratio with SimKO adjustments
+    """
+    # 1. Identify forking tokens (high-entropy)
+    entropy = compute_token_entropy(pred, mask)
+    valid_entropy = entropy[mask.bool()]
+    if len(valid_entropy) > 0:
+        tau = torch.quantile(valid_entropy, tau_percentile / 100.0)
+        w = (entropy > tau).float()
+    else:
+        w = torch.zeros_like(mask)
+    
+    # 2. Get top-K tokens and compute their ratios
+    _, topk_indices = torch.topk(pred, k=K, dim=-1)
+    new_log_probs_full = torch.nn.functional.log_softmax(pred, dim=-1)
```

Comment:
this uses up a lot of memory

Reply:
well this needs grads, since the grads are flowing through this softmax, we need to combine the way we're currently getting logprobs with the top-k logprob method, as the logprobs we need for the GRPO/GSPO loss are not guaranteed to be within the top-k. the quick and dirty way may be to wrap it in a checkpoint and chunk it, that way we "save" the memory by having it re-execute on backward in manageable chunks, even if it's not optimal, it should still just be a tiny portion of the compute so we can add <figuring out a way to reuse the whole logit space> to the backlog as i'm mostly worried about memory 😎